### PR TITLE
Modify Content-Security-Policy for Swagger UI

### PR DESCRIPTION
**Pull request by KristianGrafana**
**What is this feature?**

This PR adds a script nonce to the Swagger UI and dynamically adds `https://unpkg.com` as an allowed source in the `style-src` Content-Security-Policy (CSP) attribute in Grafana.

**Why do we need this feature?**

Currently it is not possible to use Swagger UI together with the strict CSP defined in Grafana. It is possible to modify the CSP to allow resources for Swagger UI to be loaded, but that would affect the whole Grafana instance and decrease the overall protection that the strict CSP currently offers. 

This PR does not modify the CSP other than dynamically for Swagger UI. With this fix we do not need to allow `https://unpkg.com` globally.

**Who is this feature for?**

Users that want to use the Swagger UI and a Content-Security-Policy at the same time.

**Which issue(s) does this PR fix?**:

None created.

### CODE REVIEW

File: pkg/api/openapi3.go [hunk 1]
```diff
@@ -2,10 +2,20 @@ package api
...
 func openapi3(c *contextmodel.ReqContext) {
-	c.HTML(http.StatusOK, "openapi3", nil)
+	data := map[string]interface{}{
+		"Nonce": c.RequestNonce,
+	}
+
+	// Add CSP for unpkg.com to allow loading of Swagger UI assets
+	existingCSP := c.Resp.Header().Get("Content-Security-Policy")
+	newCSP := strings.Replace(existingCSP, "style-src", "style-src https://unpkg.com/", 1)
+	c.Resp.Header().Set("Content-Security-Policy", newCSP)
```
Comment (jcalisto) [inline pkg/api/openapi3.go:18]: If there is no CSP defined, Header().Get() will return "" and we'll end up setting an empty CSP. I don't think that's a problem, but maybe we can avoid it?

```suggestion
	if existingCSP := c.Resp.Header().Get("Content-Security-Policy"); existingCSP != "" {
		newCSP := strings.Replace(existingCSP, "style-src", "style-src https://unpkg.com/", 1)
		c.Resp.Header().Set("Content-Security-Policy", newCSP)
	}
```

**Review by jcalisto — APPROVED** [review]

**Comment by samjewell** [comment]
I was wondering what's blocking this work?

Our [docs](https://grafana.com/docs/grafana/latest/developers/http_api/) still say:
> Users can browser and try out both via the Swagger UI editor (served by the grafana server) by navigating to /swagger-ui and /openapi3 respectively.

**Comment by KristianGrafana** [comment]
@samjewell I'm not allowed to merge this into the main branch; an authorized user needs to do that. Please let me know who we could ping in order to get this reviewed and merged.

**Review by papagian — CHANGES_REQUESTED** [review]
If I start the server locally with `make run` I can no more access the UI: it fails with the following error:
```
ERROR[08-01|10:00:29] Request error                            logger=context userId=0 orgId=0 uname= error="Context.HTML:template: index:4:14: executing \"index\" at <.CSPEnabled>: can't evaluate field CSPEnabled in type *dtos.IndexViewData" stack="/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/context.go:106 (0x105a2abdb)\n\t(*Context).HTML: panic(\"Context.HTML:\" + err.Error())\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/api/login.go:140 (0x106a41957)\n\t(*HTTPServer).LoginView: c.HTML(http.StatusOK, getViewIndex(), viewData)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/api/response/web_hack.go:40 (0x105fd4f67)\n\twrap_handler.func3: handle(reqCtx(w, r))\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:134 (0x105a2bd53)\n\tmwFromHandler.func1.1: h.ServeHTTP(mrw, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/services/contexthandler/contexthandler.go:220 (0x1061437f7)\n\t(*ContextHandler).Middleware.func1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/render.go:44 (0x105a2c8d3)\n\tRenderer.func1.1: next.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/csrf/csrf.go:64 (0x106856fbf)\n\t(*CSRF).Middleware.func1.1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/recovery.go:171 (0x10614faab)\n\tRecovery.func1.1: next.ServeHTTP(w, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/logger.go:42 (0x10614d347)\n\tLogger.func1.1: next.ServeHTTP(rw, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/request_metrics.go:59 (0x1061507ef)\n\tRequestMetrics.func1.1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/request_tracing.go:88 (0x106151edf)\n\tRequestTracing.func1.1: next.ServeHTTP(w, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/context.go:49 (0x105a2a723)\n\t(*Context).run: h.ServeHTTP(ctx.Resp, ctx.Req)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/router.go:155 (0x105a2dcfb)\n\t(*Router).Handle.func1: c.run()\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/router.go:211 (0x105a2e7af)\n\t(*Router).ServeHTTP: leaf.handle(rw, req, nil)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:166 (0x105a2c22b)\n\t(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2936 (0x1050ad3b7)\n\tserverHandler.ServeHTTP: handler.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:1995 (0x1050a94ff)\n\t(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)\n/opt/homebrew/opt/go/libexec/src/runtime/asm_arm64.s:1172 (0x104dc2c83)\n\tgoexit: MOVD\tR0, R0\t// NOP\n"
```

BTW we also have an issue that the swagger is no more accessible in the cloud #69757
(I remember that it used to work in the past but I was able to reproduce it myself with my own instnace)
I wonder if this is anyhow related and possible would address it?

**Comment by papagian** [comment]
> If I start the server locally with `make run` ~I can no more access the UI: it fails with the following error~:
> 
> ```
> ERROR[08-01|10:00:29] Request error                            logger=context userId=0 orgId=0 uname= error="Context.HTML:template: index:4:14: executing \"index\" at <.CSPEnabled>: can't evaluate field CSPEnabled in type *dtos.IndexViewData" stack="/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/context.go:106 (0x105a2abdb)\n\t(*Context).HTML: panic(\"Context.HTML:\" + err.Error())\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/api/login.go:140 (0x106a41957)\n\t(*HTTPServer).LoginView: c.HTML(http.StatusOK, getViewIndex(), viewData)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/api/response/web_hack.go:40 (0x105fd4f67)\n\twrap_handler.func3: handle(reqCtx(w, r))\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:134 (0x105a2bd53)\n\tmwFromHandler.func1.1: h.ServeHTTP(mrw, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/services/contexthandler/contexthandler.go:220 (0x1061437f7)\n\t(*ContextHandler).Middleware.func1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/render.go:44 (0x105a2c8d3)\n\tRenderer.func1.1: next.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:140 (0x105a2bdf7)\n\tmwFromHandler.func1.1: next.ServeHTTP(ctx.Resp, ctx.Req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/csrf/csrf.go:64 (0x106856fbf)\n\t(*CSRF).Middleware.func1.1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/recovery.go:171 (0x10614faab)\n\tRecovery.func1.1: next.ServeHTTP(w, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/logger.go:42 (0x10614d347)\n\tLogger.func1.1: next.ServeHTTP(rw, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/request_metrics.go:59 (0x1061507ef)\n\tRequestMetrics.func1.1: next.ServeHTTP(w, r)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/middleware/request_tracing.go:88 (0x106151edf)\n\tRequestTracing.func1.1: next.ServeHTTP(w, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2122 (0x1050aa717)\n\tHandlerFunc.ServeHTTP: f(w, r)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/context.go:49 (0x105a2a723)\n\t(*Context).run: h.ServeHTTP(ctx.Resp, ctx.Req)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/router.go:155 (0x105a2dcfb)\n\t(*Router).Handle.func1: c.run()\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/router.go:211 (0x105a2e7af)\n\t(*Router).ServeHTTP: leaf.handle(rw, req, nil)\n/Users/josefk/go/src/github.com/grafana/grafana/pkg/web/macaron.go:166 (0x105a2c22b)\n\t(*Macaron).ServeHTTP: m.Router.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:2936 (0x1050ad3b7)\n\tserverHandler.ServeHTTP: handler.ServeHTTP(rw, req)\n/opt/homebrew/opt/go/libexec/src/net/http/server.go:1995 (0x1050a94ff)\n\t(*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)\n/opt/homebrew/opt/go/libexec/src/runtime/asm_arm64.s:1172 (0x104dc2c83)\n\tgoexit: MOVD\tR0, R0\t// NOP\n"
> ```

my mistake: I had to rebuilt the frontend

**Review by papagian — APPROVED** [review]

**Comment by papagian** [comment]
I have noticed that this is quite old - the `swagger-ui` is now enabled by default.
I would worth syncing with the latest main before merging

**Comment by KristianGrafana** [comment]
@papagian Thank you! Feel free to sync this and see if that solves the test fails :) And yes, this currently breaks SwaggerUI and OpenAPI in Hosted Grafana, and this PR will (hopefully) fix it. It's not a perfect solution, but if we can get this merged we can start drafting on a more sustainable solution (for the whole codebase).

**Comment by KristianGrafana** [comment]
Hey @papagian - is there an issue with deploying this to hg? If so, I can provide with another fix.

**Pull request merged**